[Kogito-8164] Kogito Serverless Workflow Operator Documentation (#31)
davidesalerno/kogito-serverless-operator: CONTRIBUTING.md

    Contribute actively to development (see the section below)

    Use it and report any feedback, improvement or bug you may find via Github, mailing list or chat.

    Contribute by writing missing documentation or blog posts about the features around Kogito

    Tweet, like and socialize Kogito in your preferred social network

    Enjoy the talks that the contributors submit in various conferences around the world

# Contributing to the Kogito Serverless Operator codebase

The main project is written in go. 
Kogito Serverless Workflow Operator is built on top of Kubernetes through Custom Resource Definitions.

- Workflow
- Platform
- Build

This project aims to follow the Kubernetes [Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/)

It uses [Controllers](https://kubernetes.io/docs/concepts/architecture/controller/)
which provides a reconcile function responsible for synchronizing resources untile the desired state is reached on the cluster

# Contributing to the Kogito Serverless Operator

# PreRequisites

> The Operator's controllers and the configurations are generated using the Operator sdk, the code is written in Go, the tasks are executed using a Makefile

Author: ok

In order to build the project, you need to comply with the following requirements:

> operator-sdk-v1.24.0 (https://sdk.operatorframework.io/docs/building-operators/golang/installation/)

Reviewer: use bullet points and enforce the minimum version requirement, e.g. operator sdk 1.24+

Author: ok

> More information can be found via the [Kubebuilder Documentation](https://book.kubebuilder.io/introduction.html)

Reviewer: mode information about what? is it a requirement?

Author: No, moved

> Go 1.19 (https://go.dev/dl/)

Reviewer: hide links

Author: ok

GNU Make: 

> used to define composite build actions. This should be already installed or available as a package if you have a good OS (https://www.gnu.org/software/make/).

Author: :) https://camel.apache.org/camel-k/next/contributing/developers.html

> used to define composite build actions. This should be already installed or available as a package if you have a good OS (https://www.gnu.org/software/make/).

Reviewer: imho, we shouldn't use it.

> used to define composite build actions. This should be already installed or available as a package if you have a good OS (https://www.gnu.org/software/make/).

Author: Removed, but was to show the root of the expression

**NOTE:** Run `make help` for more information on all potential `make` targets

## Getting Started
You’ll need a Kubernetes cluster to run against. You can use 
- [KIND](https://sigs.k8s.io/kind) 
- [MINIKUBE](https://minikube.sigs.k8s.io)  

> - [CRC](https://console.redhat.com/openshift/create/local) 

Reviewer: does it worth to mention the OpenShift free tier?

> to get a local cluster for testing, or run against a remote cluster.

Reviewer: ```suggestion
```

Author: ok

> **Note:** Your controller will automatically use the current context in your kubeconfig file (i.e. whatever cluster `kubectl cluster-info` shows).

Reviewer 2: +1

Even if I don't understand the note. It depends where you will execute the controller:
- if you execute it locally it's true
- if you deploy it in a Pod it will use the context for that Pod

Since we are in the Getting started section I don't think that the note it is really meaningful here.

It could be useful instead in the [Test it out](https://github.com/davidesalerno/kogito-serverless-operator/tree/98defe6f1333627982775ffc4dd3363134114f55#test-it-out) section

### Building

```sh
make container-build
```

### Modifying the API definitions
If you are editing the API definitions, generate the manifests such as CRs or CRDs using:

```sh
make manifests
```

> ### Running on the cluster

Reviewer: what about instead adding this, mention only the operator-sdk option to run it?
this steps should be handled by olm, it is better to exemplify the operator-sdk option together the build process and olm deploy.

Reviewer 2: @spolti This section is generated by the operator-sdk so it could be probably useful to easily deploy the operator in a local development cluster.

I would standardize the commands from docker-* to container-*

Reviewer: ok, but I still thinking that it worth to mention how to use operator-sdk to run it for local development/testing.

Author: Section removed and leaved only on the README

1. Install Instances of Custom Resources:

```sh
kubectl apply -f config/samples/
```

2. Build and push your image to the location specified by `IMG`:

```sh
make docker-build docker-push IMG=<some-registry>/kogito-serverless-operator:tag
```

3. Deploy the controller to the cluster with the image specified by `IMG`:

```sh
make deploy IMG=<some-registry>/kogito-serverless-operator:tag
```

### Uninstall CRDs
To delete the CRDs from the cluster:

```sh
make uninstall
```

### Undeploy controller
UnDeploy the controller to the cluster:

```sh
make undeploy
```

> Minikube or CRC

Reviewer: what does it means?

Author: wrong c&p

## Test the Greeting workflow on Minikube

A good starting point to check that everything is working well, it is the [Greeting workflow](https://github.com/kiegroup/kogito-examples/blob/stable/README.md#serverless-workflow-getting-started).

> Follow these steps to create a container that you can than deploy as a Service on Kubernetes or KNative.

Reviewer: make clear that for this example you will use minikube.

Reviewer 2: +1

Author: Title "Test the Greeting workflow on Minikube" isn't explicit enough ? :D https://github.com/davidesalerno/kogito-serverless-operator/pull/31/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055R99

1. Start Minikube
```sh 

> minikube start --cpus 4 --memory 4096 --addons registry --insecure-registry "10.0.0.0/24"

Reviewer 2: @spolti The insecure registry option is probably meant only to disable SSL communication between the registry and the clients (kaniko or the kube control plane) https://minikube.sigs.k8s.io/docs/handbook/registry/#enabling-insecure-registries in order to avoid the load of the CA certificates into the kaniko image.

When we will have the Platform we will be able to configure also a secure communication but at the moment no.

```
2. Create a namespace for the building phase

```sh
kubectl create namespace kogito-builder
```

3. Create a secret
```sh
kubectl create secret docker-registry regcred --docker-server=<registry_url> --docker-username=<registry_username> --docker-password=<registry_password> --docker-email=<registry_email> -n kogito-builder
```

4. Build and push your image to the location specified by `IMG`:

```sh

> make container-build container-push IMG=<some-registry>/kogito-serverless-operator:tag

Reviewer: isn't these steps the same than the ones you've added before?

Reviewer 2: +1

In general I'll try to avoid duplication between the README and this file.

I'm expecting to find here how to contribute to the code and some useful tips for the development that are not included in the general README.

Author: Added a link to the readme about the Running on the cluster

```

5. Deploy the controller to the cluster with the image specified by `IMG`:

```sh
make deploy IMG=<some-registry>/kogito-serverless-operator:tag
```

6. Create a dedicated Namespace for the test:

```sh

> kubectl create namespace greeting-workflow

Author: Removed and added a link to the README.md

```

7. Install Instances of Custom Resources:

```sh
kubectl apply -f config/samples/sw.kogito.kie.org__v08_kogitoserverlessworkflow.yaml -n greeting-workflow
```

8. Default configuration

> A configmap called kogito-serverless-operator-builder-config will be created under the kogito-serverless-operator-system

Reviewer: from where this namespace is coming from?

Author: IS created by the operator, but we add the configmap under kogito-builder namespace

Author: ok

> namespace when the Operator will be installer, and it contains:

Author: yep

- DEFAULT_BUILDER_RESOURCE = kogito_builder_dockerfile.yaml
- DEFAULT_WORKFLOW_DEXTENSION = .sw.json
- DEFAULT_KANIKO_SECRET_DEFAULT = regcred

> - DEFAULT_REGISTRY_REPO = quay.io/kiegroup

Reviewer: for minikube let's use dev.local.

Reviewer 2: @desmax74 I'll do a test with dev.local and if it works on both sides (for pushing but also for pulling in the deployment phase) we can use it.

> - kogito_builder_dockerfile.yaml = dockerfile content

Reviewer: <Dockerfile_content>

Author: ok

For the local development the DEFAULT_REGISTRY_REPO must be changed

# Customize Builder Image

> At the startup a 

Author: fixed

[Dockerfile](./config/manager/kogito_builder_dockerfile.yaml) 
is placed in a configmap

> This Dockerfile use a base image called [swfbuilder](https://github.com/kiegroup/kogito-images/tree/master/modules/kogito-swf-builder) with installed:

Reviewer: main

Author: ?

- openjdk 11

> - maven

Reviewer: which version?

Author: added

> A Quarkus project  /home/kogito/serverless-workflow-project with the extensions

Reviewer: ```suggestion
A Quarkus project `/home/kogito/serverless-workflow-project` with the extensions
```

Author: fixed

- quarkus-kubernetes 
- kogito-quarkus-serverless-workflow 
- kogito-addons-quarkus-knative-eventing

> all the dependencies needed in the /home/kogito/.m2

Reviewer: ```suggestion
All the dependencies are store in the `/home/kogito/.m2` directory.
```
but I guess it is not needed

Author: fixed

> additional scripts in case of need to: 

Reviewer: in case of need to, what exactly?
can you please ellaborate?

Author: ok

- add other quarkus extensions,
- build the project after adding other files/jaav classes
- create a new project

> You can customize your final Image using this Dockerfile

Reviewer: which Dockerfile?
